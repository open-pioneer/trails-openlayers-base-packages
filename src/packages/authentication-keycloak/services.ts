// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import Keycloak, {
    KeycloakInitOptions,
    KeycloakLoginOptions,
    KeycloakLogoutOptions
} from "keycloak-js";
import { KeycloakConfigProvider } from "./api";
/*export const keycloak = new Keycloak({
    url: "https://auth.ldproxy.net/",
    realm: "ii",
    clientId: "it-nrw"
});

export const keycloak = new Keycloak({
    url: "https://hsi-pex0-13620.service.it.nrw.de/keycloak/",
    realm: "trails",
    clientId: "trails-sample"
});
*/
let uKeycloak = new Keycloak();

export class KeycloakConfigProviderImpl implements KeycloakConfigProvider {
    setKeycloak(keycloak: Keycloak) {
        uKeycloak = keycloak;
        return keycloak;
    }

    getInitOptions(
        options: KeycloakInitOptions | undefined = {
            onLoad: "check-sso",
            pkceMethod: "S256",
            scope: "data:read"
        }
    ) {
        return options;
    }

    getLoginOptions(options: KeycloakLoginOptions | undefined = { redirectUri: undefined }) {
        return options;
    }

    getLogoutOptions(options: KeycloakLogoutOptions | undefined = { redirectUri: undefined }) {
        return options;
    }

    getRefreshOptions(
        options: { autoRefresh: boolean; interval: number; timeLeft: number } | undefined = {
            autoRefresh: true,
            interval: 6000,
            timeLeft: 70
        }
    ) {
        return options;
    }
}

export { uKeycloak };
