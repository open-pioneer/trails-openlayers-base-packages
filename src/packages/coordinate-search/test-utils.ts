// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { act, fireEvent } from "@testing-library/react";

export function showDropdown(projSelect: HTMLElement) {
    // open dropdown to include options in snapshot; react-select creates list of options in dom after opening selection
    act(() => {
        fireEvent.keyDown(projSelect, { key: "ArrowDown" });
    });
}

export function getCurrentOptions() {
    return Array.from(
        document.getElementsByClassName("coordinate-input-select__option")
    ) as HTMLElement[];
}

export function getCurrentOptionValues(options: HTMLElement[]) {
    const values: (string | null)[] = [];
    for (const opt of options) {
        values.push(opt.textContent);
    }
    return values;
}

export function getClearButton(coordinateSearchGroup: Element) {
    const buttonDiv = coordinateSearchGroup.querySelector(".chakra-input__right-element");
    if (!buttonDiv) {
        throw new Error("buttons not rendered");
    }
    const clearButton = buttonDiv.querySelector(".coordinate-input-clear-button");
    if (!clearButton) {
        throw new Error("clear button not rendered");
    }
    return clearButton;
}

export function getCopyButton(coordinateSearchGroup: Element) {
    const buttonDiv = coordinateSearchGroup.querySelector(".chakra-input__right-element");
    if (!buttonDiv) {
        throw new Error("buttons not rendered");
    }
    const copyButton = buttonDiv.querySelector(".coordinate-input-copy-button");
    if (!copyButton) {
        throw new Error("copy button not rendered");
    }
    return copyButton;
}
