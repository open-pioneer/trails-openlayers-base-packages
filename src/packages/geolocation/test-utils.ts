// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import OlMap from "ol/Map";
import { Circle, Fill, Stroke, Style } from "ol/style";
import { vi } from "vitest";
import { GeolocationController } from "./GeolocationController";

export function mockSuccessGeolocation(coords: number[]) {
    vi.spyOn(navigator, "geolocation", "get").mockReturnValue({
        clearWatch() {},
        getCurrentPosition() {},
        watchPosition(success) {
            setTimeout(() => {
                success({
                    coords: {
                        latitude: coords[0],
                        longitude: coords[1],
                        accuracy: 2500,
                        altitude: null,
                        altitudeAccuracy: null,
                        heading: null,
                        speed: null
                    }
                } as GeolocationPosition);
            }, 1);

            return 123;
        }
    } satisfies Partial<Geolocation>);
}

export function mockErrorGeolocation() {
    vi.spyOn(navigator, "geolocation", "get").mockReturnValue({
        clearWatch() {},
        getCurrentPosition() {},
        watchPosition(success, error) {
            setTimeout(() => {
                error?.({
                    code: 2,
                    message: "POSITION_UNAVAILABLE",
                    PERMISSION_DENIED: 1,
                    POSITION_UNAVAILABLE: 2,
                    TIMEOUT: 3
                });
            }, 1);

            return 123;
        }
    } satisfies Partial<Geolocation>);
}

export function setup() {
    const olMap = new OlMap();
    return {
        olMap,
        controller: new GeolocationController(olMap, (error) => {
            console.error("Unexpected error", error);
        })
    };
}

export function setupWithCustomProperties() {
    const olMap = new OlMap();
    const maxZoom: number = getCustomMaxZoom();
    const positionFeatureStyle: Style = getCustomPositionStyle();
    const accuracyFeatureStyle: Style = getCustomAccuracyStyle();
    const trackingOptions: PositionOptions = getCustomTrackingOptions();

    const controller = new GeolocationController(
        olMap,
        (error) => {
            console.error("Unexpected error", error);
        },
        trackingOptions
    );
    controller.setMaxZoom(maxZoom);
    controller.setPositionFeatureStyle(positionFeatureStyle);
    controller.setAccuracyFeatureStyle(accuracyFeatureStyle);
    return {
        olMap,
        controller
    };
}

export function getCustomMaxZoom() {
    return 20;
}
export function getCustomPositionStyle() {
    return new Style({
        image: new Circle({
            fill: new Fill({
                color: "rgba(255,255,255,0.4)"
            }),
            stroke: new Stroke({
                color: "#3399CC",
                width: 1.25
            }),
            radius: 5
        }),
        fill: new Fill({
            color: "rgba(255,255,255,0.4)"
        }),
        stroke: new Stroke({
            color: "#3399CC",
            width: 1.25
        })
    });
}

export function getCustomAccuracyStyle() {
    return new Style({
        image: new Circle({
            fill: new Fill({
                color: "rgba(255,255,255,0.4)"
            }),
            stroke: new Stroke({
                color: "#3399CC",
                width: 1.25
            }),
            radius: 5
        }),
        fill: new Fill({
            color: "rgba(255,255,255,0.4)"
        }),
        stroke: new Stroke({
            color: "#3399CC",
            width: 1.25
        })
    });
}

function getCustomTrackingOptions() {
    return {
        "enableHighAccuracy": true,
        "timeout": 20,
        "maximumAge": 50
    };
}
