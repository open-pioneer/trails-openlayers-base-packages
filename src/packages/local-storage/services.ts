// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
export { LocalStorageServiceImpl } from "./LocalStorageServiceImpl";
