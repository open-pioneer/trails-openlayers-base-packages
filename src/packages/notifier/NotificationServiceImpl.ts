// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import type { NotificationLevel, NotificationOptions, NotificationService } from "./api";
import { Resource, createLogger } from "@open-pioneer/core";
import type { ReactNode } from "react";
const LOG = createLogger("notifier:NotificationService");

export interface Notification {
    title: ReactNode;
    message: ReactNode;
    level: NotificationLevel;
    displayDuration: number | undefined;
}

/**
 * The notification handler is implemented by the UI in order to receive events.
 * There can only be one handler at a time.
 *
 * Calls to the handler may be buffered to allow for notifications before the UI has become ready.
 */
export interface NotificationHandler {
    /** Called by the service to emit the notification. */
    showNotification(notification: Notification): void;

    /** Removes all active notifications. */
    closeAll(): void;
}

export interface InternalNotificationAPI extends NotificationService {
    /**
     * Registers the notification handler.
     * The returned resource can be used to unregister it again.
     */
    registerHandler(handler: NotificationHandler): Resource;
}

export class NotificationServiceImpl implements InternalNotificationAPI {
    #handler: NotificationHandler | undefined;
    #buffered: [keyof NotificationHandler, ...unknown[]][] | undefined;
    #checkTimeoutId: any; // eslint-disable-line @typescript-eslint/no-explicit-any

    constructor() {
        if (import.meta.env.DEV) {
            this.#checkTimeoutId = setTimeout(() => {
                this.#checkHandlerRegistration();
                this.#checkTimeoutId = undefined;
            }, 5000);
        }
    }

    destroy() {
        clearTimeout(this.#checkTimeoutId);
        this.#checkTimeoutId = undefined;
    }

    notify(options: NotificationOptions): void {
        this.#dispatchHandlerMethod("showNotification", {
            title: options.title ?? undefined,
            message: options.message ?? undefined,
            level: options.level ?? "info",
            displayDuration: options.displayDuration
        });
    }

    closeAll(): void {
        this.#dispatchHandlerMethod("closeAll");
    }

    registerHandler(handler: NotificationHandler): Resource {
        // We only support exactly one handler.
        if (this.#handler) {
            LOG.warn(
                "A notification handler has already been registered; this new handler will be ignored.\n" +
                    "The <Notifier /> component has likely been used twice in your application."
            );
            return {
                destroy() {
                    return undefined;
                }
            };
        }

        // Dispatch buffered calls (if any).
        // These happened during the time no handler was registered.
        this.#handler = handler;
        const buffered = this.#buffered;
        this.#buffered = undefined;
        if (buffered) {
            for (const [name, ...args] of buffered) {
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                (handler[name] as any)(...args);
            }
        }

        // Return a resource to undo the registration
        let destroyed = false;
        return {
            destroy: () => {
                if (destroyed) {
                    return;
                }

                destroyed = true;
                if (this.#handler === handler) {
                    this.#handler = undefined;
                }
            }
        };
    }

    /**
     * Calls a method on the handler (if present) or buffers the method call for later once the handler
     * has arrived.
     */
    #dispatchHandlerMethod<Method extends keyof NotificationHandler>(
        method: Method,
        ...args: Parameters<NotificationHandler[Method]>
    ): void;
    #dispatchHandlerMethod(method: keyof NotificationHandler, ...args: unknown[]): void {
        if (this.#handler) {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (this.#handler[method] as any)(...args);
        } else {
            const buffered = (this.#buffered ??= []);
            if (buffered.length >= 1024) {
                LOG.error(
                    "Internal notification buffer overflow: this event will be dropped to prevent a memory leak.\n" +
                        "Make sure that the UI is configured to display notifications (use <Notifier />).",
                    { method, args }
                );
                return;
            }
            buffered.push([method, ...args]);
        }
    }

    #checkHandlerRegistration() {
        if (!this.#handler) {
            LOG.warn(
                `No notification handler has been registered: notifications will not be visible.\n` +
                    `Make sure that your app contains the <Notifier /> component.`
            );
        }
    }
}
