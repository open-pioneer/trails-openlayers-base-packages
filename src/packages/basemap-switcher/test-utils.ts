// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { MapConfig, MapConfigProvider, MapRegistry } from "@open-pioneer/map";
import TileLayer from "ol/layer/Tile";
import OSM from "ol/source/OSM";
import Stamen from "ol/source/Stamen";
import { createService } from "@open-pioneer/test-utils/services";
import { MapRegistryImpl } from "@open-pioneer/map/MapRegistryImpl";
import { PackageContextProviderProps } from "@open-pioneer/test-utils/react";
import { screen, waitFor } from "@testing-library/react";

// used to avoid a "ResizeObserver is not defined" error
import ResizeObserver from "resize-observer-polyfill";
global.ResizeObserver = ResizeObserver;

export interface SimpleMapOptions {
    center?: { x: number; y: number };
    zoom?: number;
}

export const MAP_ID = "test";

export async function waitForMapMount(parentTestId = "base") {
    return await waitFor(async () => {
        const domElement = await screen.findByTestId(parentTestId);
        const container = domElement.querySelector(".ol-viewport");
        if (!container) {
            throw new Error("map not mounted");
        }
        return domElement;
    });
}
export async function setupMap(options?: SimpleMapOptions) {
    const mapId = "test";
    const mapConfig: MapConfig = {
        initialView: {
            kind: "position",
            center: options?.center ?? { x: 847541, y: 6793584 },
            zoom: options?.zoom ?? 10
        },
        projection: "EPSG:3857",
        layers: [
            {
                id: "b-1",
                title: "OSM",
                isBaseLayer: true,
                visible: false,
                layer: new TileLayer({
                    source: new OSM()
                })
            },
            {
                id: "b-2",
                title: "Toner",
                isBaseLayer: true,
                visible: true,
                layer: new TileLayer({
                    source: new Stamen({ layer: "toner" })
                })
            }
        ]
    };

    const registry = await createService(MapRegistryImpl, {
        references: {
            providers: [new MapConfigProviderImpl(mapId, mapConfig)]
        }
    });

    return { mapId, registry };
}

export function createPackageContextProviderProps(
    service: MapRegistry
): PackageContextProviderProps {
    return {
        services: {
            "map.MapRegistry": service
        }
    };
}

class MapConfigProviderImpl implements MapConfigProvider {
    mapId = "default";
    mapConfig: MapConfig;

    constructor(mapId: string, mapConfig?: MapConfig | undefined) {
        this.mapId = mapId;
        this.mapConfig = mapConfig ?? {};
    }

    getMapConfig(): Promise<MapConfig> {
        return Promise.resolve(this.mapConfig);
    }
}
