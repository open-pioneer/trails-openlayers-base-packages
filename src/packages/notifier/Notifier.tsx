// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { useToast } from "@open-pioneer/chakra-integration";
import { useEvent } from "@open-pioneer/react-utils";
import { useService } from "open-pioneer:react-hooks";
import { ReactElement, useEffect, useState } from "react";
import { InternalNotificationAPI, Notification } from "./NotificationServiceImpl";
import { WarningTwoIcon } from "@chakra-ui/icons";

/** Props supported by the {@link Notifier} component. */
export interface NotifierProps {
    /**
     * The position for new notifications.
     *
     * @default "top-right"
     */
    position?: "top" | "top-left" | "top-right" | "bottom" | "bottom-left" | "bottom-right";
}

const isDev = import.meta.env.DEV;

/**
 * Shows notifications sent via the `NotificationService`.
 *
 * Only one instance of `<Notifier />` should be present in the application.
 * It currently does not matter where the Notifier is located in the react tree.
 *
 * ```ts
 * import { Notifier } from "@open-pioneer/notifier";
 *
 * export function AppUI() {
 *     return (
 *         <>
 *             <Notifier position="top-right" />
 *             <h1>Your application</h1>
 *         </>
 *     );
 * }
 * ```
 */
export function Notifier(props: NotifierProps): ReactElement {
    const { position = "top-right" } = props;
    const toast = useToast();
    const notifications = useService("notifier.NotificationService") as InternalNotificationAPI;
    const [ready, setReady] = useState(!isDev);

    const showNotification = useEvent((notification: Notification) => {
        // use different icons for warning and error
        const icon =
            notification.level === "error" ? <WarningTwoIcon h="100%" w="100%" /> : undefined;
        toast({
            position,
            title: notification.title,
            description: notification.message || null,
            status: notification.level,
            isClosable: true,
            duration: notification.displayDuration ?? null, // Null: does not auto-close
            icon
        });
    });
    const closeAll = useEvent(() => {
        toast.closeAll();
    });

    useEffect(() => {
        /*
         * Delay registering the notification handler a bit during development.
         * Chakra's toast implementation clears its store when it gets unmounted.
         * During development, that may also happen during initialization because of React's
         * StrictMode (https://react.dev/reference/react/StrictMode).
         *
         * Unfortunately, that means that toasts that were emitted very early (such as in service constructors)
         * will be deleted. Waiting for a few milliseconds prevents this issue and has no effect on production.
         */
        if (isDev && !ready) {
            const id = setTimeout(() => setReady(true), 100);
            return () => clearTimeout(id);
        }

        const handle = notifications.registerHandler({
            showNotification,
            closeAll
        });
        return () => handle.destroy();
    }, [ready, notifications, showNotification, closeAll]);

    // No actual UI representation right now
    return <></>;
}
