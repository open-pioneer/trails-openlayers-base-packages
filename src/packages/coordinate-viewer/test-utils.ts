// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { OlMapConfigurationProvider } from "@open-pioneer/experimental-ol-map";
import { OlMapRegistry } from "@open-pioneer/experimental-ol-map/services";
import { ServiceOptions } from "@open-pioneer/runtime";
import { PackageContextProviderProps } from "@open-pioneer/test-utils/react";
import { createService } from "@open-pioneer/test-utils/services";
import { waitFor, screen } from "@testing-library/react";
import { MapOptions } from "ol/Map";
import View from "ol/View";
import TileLayer from "ol/layer/Tile";
import OSM from "ol/source/OSM";

// used to avoid a "ResizeObserver is not defined" error
import ResizeObserver from "resize-observer-polyfill";
global.ResizeObserver = ResizeObserver;

export interface SimpleMapOptions {
    center?: [number, number];
    zoom?: number;
}

export const MAP_ID = "test";

export async function waitForMapMount(parentTestId = "base") {
    return await waitFor(async () => {
        const domElement = await screen.findByTestId(parentTestId);
        const container = domElement.querySelector(".ol-viewport");
        if (!container) {
            throw new Error("map not mounted");
        }
        return domElement;
    });
}

export async function setupMap(options?: SimpleMapOptions) {
    const mapOptions: MapOptions = {
        view: new View({
            projection: "EPSG:3857",
            center: options?.center ?? [847541, 6793584],
            zoom: options?.zoom ?? 10
        }),
        layers: [
            new TileLayer({
                source: new OSM(),
                properties: { title: "OSM" }
            })
        ]
    };

    const mapConfigProvider = await createService(MapConfigProvider, {
        properties: {
            mapOptions: mapOptions,
            mapId: MAP_ID
        }
    });
    const registry = await createService(OlMapRegistry, {
        references: {
            providers: [mapConfigProvider]
        }
    });

    return { mapId: MAP_ID, registry };
}

export function createPackageContextProviderProps(
    service: OlMapRegistry
): PackageContextProviderProps {
    return {
        services: {
            "ol-map.MapRegistry": service
        }
    };
}

class MapConfigProvider implements OlMapConfigurationProvider {
    mapId = "default";
    mapOptions: MapOptions = {};

    constructor(options: ServiceOptions) {
        if (options.properties.mapOptions) {
            this.mapOptions = options.properties.mapOptions as MapOptions;
        }
        if (options.properties.mapId) {
            this.mapId = options.properties.mapId as string;
        }
    }

    getMapOptions(): Promise<MapOptions> {
        return Promise.resolve(this.mapOptions);
    }
}
