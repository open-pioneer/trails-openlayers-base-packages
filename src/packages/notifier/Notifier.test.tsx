// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { expect, it } from "vitest";
import { createService } from "@open-pioneer/test-utils/services";
import { NotificationServiceImpl } from "./NotificationServiceImpl";
import { render, screen, waitForElementToBeRemoved, act, waitFor } from "@testing-library/react";
import { PackageContextProvider } from "@open-pioneer/test-utils/react";
import { Notifier } from "./Notifier";

function findToastInnerElement(messageDiv: HTMLElement) {
    let currentElement = messageDiv.parentElement;
    while (currentElement) {
        if (currentElement.classList.contains("chakra-toast__inner")) {
            return currentElement;
        } else {
            currentElement = currentElement.parentElement;
        }
    }
    return undefined;
}

it("shows notifications as toasts", async () => {
    const service = await createService(NotificationServiceImpl, {});

    const services = {
        "notifier.NotificationService": service
    };

    render(
        <PackageContextProvider services={services}>
            <Notifier />
        </PackageContextProvider>
    );

    act(() => {
        service.notify({
            title: "test",
            message: <div data-testid="notification-message">Message</div>
        });
    });

    const messageDiv = await screen.findByTestId("notification-message", {});
    const toastElement = findToastInnerElement(messageDiv);

    expect(toastElement).toMatchSnapshot();
});

it("closes all notifications", async () => {
    const service = await createService(NotificationServiceImpl, {});

    const services = {
        "notifier.NotificationService": service
    };

    render(
        <PackageContextProvider services={services}>
            <Notifier />
        </PackageContextProvider>
    );

    act(() => {
        service.notify({ title: "test1" });
        service.notify({ title: "test1" });
        service.notify({ title: "test1" });
    });

    const messageElements = await waitFor(async () => {
        const messageElements = await screen.findAllByText("test1");
        if (messageElements.length !== 3) {
            throw new Error("Did not open three notifications");
        }
        return messageElements;
    });
    expect(messageElements).toHaveLength(3);

    act(() => {
        service.closeAll();
    });
    await waitForElementToBeRemoved(messageElements);
    const messageElementsAfterClear = await screen.queryAllByText("test1");
    expect(messageElementsAfterClear).toHaveLength(0);
});
