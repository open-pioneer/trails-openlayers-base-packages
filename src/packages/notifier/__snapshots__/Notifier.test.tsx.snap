// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`shows notifications as toasts 1`] = `
<div
  aria-atomic="true"
  class="chakra-toast__inner css-dixmdy"
  data-theme="light"
  role="status"
>
  <div
    class="chakra-alert css-48wy07"
    data-status="info"
    data-theme="light"
    id="toast-1"
  >
    <span
      class="chakra-alert__icon css-14ogjxt"
      data-status="info"
      data-theme="light"
    >
      <svg
        class="chakra-icon css-qk6lof"
        data-theme="light"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12,0A12,12,0,1,0,24,12,12.013,12.013,0,0,0,12,0Zm.25,5a1.5,1.5,0,1,1-1.5,1.5A1.5,1.5,0,0,1,12.25,5ZM14.5,18.5h-4a1,1,0,0,1,0-2h.75a.25.25,0,0,0,.25-.25v-4.5a.25.25,0,0,0-.25-.25H10.5a1,1,0,0,1,0-2h1a2,2,0,0,1,2,2v4.75a.25.25,0,0,0,.25.25h.75a1,1,0,1,1,0,2Z"
          fill="currentColor"
        />
      </svg>
    </span>
    <div
      class="css-njbp03"
      data-theme="light"
    >
      <div
        class="chakra-alert__title css-tidvy5"
        data-status="info"
        data-theme="light"
        id="toast-1-title"
      >
        test
      </div>
      <div
        class="chakra-alert__desc css-161kwbg"
        data-status="info"
        data-theme="light"
        id="toast-1-description"
      >
        <div
          data-testid="notification-message"
        >
          Message
        </div>
      </div>
    </div>
    <button
      aria-label="Close"
      class="css-1pq15d"
      data-theme="light"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="chakra-icon css-onkibi"
        data-theme="light"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
</div>
`;
